crates/code_generation/src/structs.rs

```diff
@@ -0,0 +1,119 @@
+use convert_case::{Case, Casing};
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote};
+use source_analyzer::{struct_hierarchy::StructHierarchy, structs::Structs};
+
+pub fn generate_structs(structs: &Structs) -> TokenStream {
+    let derives = quote! {
+        #[derive(
+            Clone,
+            Debug,
+            Default,
+            serde::Deserialize,
+            serde::Serialize,
+            serialize_hierarchy::SerializeHierarchy,
+         )]
+    };
+    let configuration = hierarchy_to_token_stream(
+        &structs.configuration,
+        format_ident!("Configuration"),
+        &derives,
+    );
+    let cyclers = structs
+        .cyclers
+        .iter()
+        .map(|(cycler_module, cycler_structs)| {
+            let cycler_module_identifier = format_ident!("{}", cycler_module.to_case(Case::Snake));
+            let main_outputs = hierarchy_to_token_stream(
+                &cycler_structs.main_outputs,
+                format_ident!("MainOutputs"),
+                &derives,
+            );
+            let additional_outputs = hierarchy_to_token_stream(
+                &cycler_structs.additional_outputs,
+                format_ident!("AdditionalOutputs"),
+                &derives,
+            );
+            let persistent_state = hierarchy_to_token_stream(
+                &cycler_structs.persistent_state,
+                format_ident!("PersistentState"),
+                &derives,
+            );
+
+            quote! {
+                pub mod #cycler_module_identifier {
+                    #main_outputs
+                    #additional_outputs
+                    #persistent_state
+                }
+            }
+        });
+
+    quote! {
+        #configuration
+        #(#cyclers)*
+    }
+}
+
+fn hierarchy_to_token_stream(
+    hierarchy: &StructHierarchy,
+    struct_name: Ident,
+    derives: &TokenStream,
+) -> TokenStream {
```

Comment:
Why does this panic now?

Reply:
All failure cases where this function panics are caused by bugs in the source analyzer. I.e. an improper hierarchy such as a `struct` directly containing another `struct` without a `field` in between.
I think its fine to panic in these cases. I tried to reach these panics with malformed paths but never succeeded.